## Common Programmatic Assessments in pandas
### Gather

In [2]:
import pandas as pd
pd.set_option('display.expand_frame_repr', False)
pd.set_option('display.max_rows', 500)
pd.set_option('display.max_rows', 500)

In [3]:
patients = pd.read_csv('insulin-trials/patients.csv')
treatments = pd.read_csv('insulin-trials/treatments.csv')
adverse_reactions = pd.read_csv('insulin-trials/adverse_reactions.csv')

### Assess
These are the programmatic assessment methods in pandas that you will probably use most often:

* .head (DataFrame and Series)
* .tail (DataFrame and Series)
* .sample (DataFrame and Series)
* .info (DataFrame only)
* .describe (DataFrame and Series)
* .value_counts (Series only)
* Various methods of indexing and selecting data (.loc and bracket notation with/without boolean indexing, also .iloc)

Try them out below and keep their results in mind. Some will come in handy later in the lesson.

Check out the [pandas API reference](https://pandas.pydata.org/pandas-docs/stable/api.html) for detailed usage information.

Try `.head` and `.tail` on the `patients` table.

In [4]:
patients.head()

patient_id assigned_sex given_name  surname                    address               city       state  zip_code        country                                     contact  birthdate  weight  height   bmi
0           1       female        Zoe  Wellish       576 Brown Bear Drive  Rancho California  California   92390.0  United States        951-719-9170ZoeWellish@superrito.com  7/10/1976   121.7      66  19.6
1           2       female     Pamela     Hill  2370 University Hill Road          Armstrong    Illinois   61812.0  United States       PamelaSHill@cuvox.de+1 (217) 569-3204   4/3/1967   118.8      66  19.2
2           3         male        Jae   Debord      1493 Poling Farm Road               York    Nebraska   68467.0  United States            402-363-6804JaeMDebord@gustr.com  2/19/1980   177.8      71  24.8
3           4         male       Liêm     Phan        2335 Webster Street         Woodbridge          NJ    7095.0  United States  PhanBaLiem@jourrapide.com+1 (732) 636-8246  7/26/1951   220.9      70  31.7
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane             Dothan          AL   36303.0  United States             334-515-7487TimNeudorf@cuvox.de  2/18/1928   192.3      27  26.1

In [5]:
patients.tail()

patient_id assigned_sex given_name      surname                   address           city     state  zip_code        country                                          contact  birthdate  weight  height   bmi
498         499         male    Mustafa    Lindström       2530 Victoria Court   Milton Mills        ME    3852.0  United States      207-477-0579MustafaLindstrom@jourrapide.com  4/10/1959   181.1      72  24.6
499         500         male      Ruman      Bisliev  494 Clarksburg Park Road         Sedona        AZ   86341.0  United States               928-284-4492RumanBisliev@gustr.com  3/26/1948   239.6      70  34.4
500         501       female      Jinke    de Keizer         649 Nutter Street  Overland Park        MO   64110.0  United States            816-223-6007JinkedeKeizer@teleworm.us  1/13/1971   171.2      67  26.8
501         502       female    Chidalu  Onyekaozulu  3652 Boone Crockett Lane        Seattle        WA   98109.0  United States  ChidaluOnyekaozulu@jourrapide.com1 360 443 2060  2/13/1952   176.9      67  27.7
502         503         male        Pat      Gersten         2778 North Avenue           Burr  Nebraska   68324.0  United States             PatrickGersten@rhyta.com402-848-4923   5/3/1954   138.2      71  19.3

Try `.sample` on the `treatments` table. - This is a good way for programmatic assesment.

In [6]:
treatments.sample(5)

given_name    surname    auralin    novodra  hba1c_start  hba1c_end  hba1c_change
75   mackenzie      mckay          -  44u - 43u         9.87       9.48          0.39
234     haylom      nebay          -  42u - 44u         7.62       7.22          0.90
83        else   andersen          -  36u - 38u         7.98       7.60           NaN
34   alexander  mathiesen  47u - 58u          -         7.96       7.55           NaN
96   abdul-nur        isa          -  54u - 50u         7.98       7.53          0.95

Try `.info` on the `treatments` table.

In [7]:
treatments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    280 non-null    object 
 1   surname       280 non-null    object 
 2   auralin       280 non-null    object 
 3   novodra       280 non-null    object 
 4   hba1c_start   280 non-null    float64
 5   hba1c_end     280 non-null    float64
 6   hba1c_change  171 non-null    float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [8]:
treatments.describe()

hba1c_start   hba1c_end  hba1c_change
count   280.000000  280.000000    171.000000
mean      7.985929    7.589286      0.546023
std       0.568638    0.569672      0.279555
min       7.500000    7.010000      0.200000
25%       7.660000    7.270000      0.340000
50%       7.800000    7.420000      0.380000
75%       7.970000    7.570000      0.920000
max       9.950000    9.580000      0.990000

The hba1c_change with value 0.99 is not feasible, looking closely at this record. The below records indicate that there is a caculation mistake between the difference caculation of hba1c_start and hba1c_end .

In [9]:
treatments[treatments.hba1c_change==0.99]

given_name      surname auralin    novodra  hba1c_start  hba1c_end  hba1c_change
32       laura  ehrlichmann       -  43u - 40u         7.95       7.46          0.99
138    giovana        rocha       -  23u - 21u         7.87       7.38          0.99
245         wu         sung       -  47u - 48u         7.61       7.12          0.99

Try `.value_counts` on the *adverse_reaction* column of the `adverse_reactions` table.

Below is a good way for checking the **Quality** issues. Look at the fileds that have 491 vs 503 count, that is missing data.Check isnull() to determine compelteness issue.

In [10]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   patient_id    503 non-null    int64  
 1   assigned_sex  503 non-null    object 
 2   given_name    503 non-null    object 
 3   surname       503 non-null    object 
 4   address       491 non-null    object 
 5   city          491 non-null    object 
 6   state         491 non-null    object 
 7   zip_code      491 non-null    float64
 8   country       491 non-null    object 
 9   contact       491 non-null    object 
 10  birthdate     503 non-null    object 
 11  weight        503 non-null    float64
 12  height        503 non-null    int64  
 13  bmi           503 non-null    float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [11]:
patients[patients['address'].isnull()]

patient_id assigned_sex given_name      surname address city state  zip_code country contact   birthdate  weight  height   bmi
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN       NaN     NaN     NaN   8/14/1950   143.4      62  26.2
219         220         male         Mỹ        Quynh     NaN  NaN   NaN       NaN     NaN     NaN    4/9/1978   237.8      69  35.1
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN       NaN     NaN     NaN   9/23/1976   165.9      63  29.4
234         235       female    Martina    Tománková     NaN  NaN   NaN       NaN     NaN     NaN    4/7/1936   199.5      65  33.2
242         243         male       John      O'Brian     NaN  NaN   NaN       NaN     NaN     NaN   2/25/1957   205.3      74  26.4
249         250         male   Benjamin       Mehler     NaN  NaN   NaN       NaN     NaN     NaN  10/30/1951   146.5      69  21.6
257         258         male        Jin         Kung     NaN  NaN   NaN       NaN     NaN     NaN   5/17/1995   231.7      69  34.2
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN       NaN     NaN     NaN   11/3/1989   158.6      63  28.1
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN       NaN     NaN     NaN   10/9/1937   175.2      61  33.1
278         279       female   Generosa        Cabán     NaN  NaN   NaN       NaN     NaN     NaN  12/16/1962   124.3      69  18.4
286         287         male      Lewis         Webb     NaN  NaN   NaN       NaN     NaN     NaN    4/1/1979   155.3      68  23.6
296         297       female        Chỉ          Lâm     NaN  NaN   NaN       NaN     NaN     NaN   5/14/1990   181.1      63  32.1

There could be a simple file transfer error that could have resulted in nulls in the address. And, there are so many other ways that nulls would have percolated in some addresses.

In [10]:
patients.describe()

patient_id      zip_code      weight      height         bmi
count  503.000000    491.000000  503.000000  503.000000  503.000000
mean   252.000000  49084.118126  173.434990   66.634195   27.483897
std    145.347859  30265.807442   33.916741    4.411297    5.276438
min      1.000000   1002.000000   48.800000   27.000000   17.100000
25%    126.500000  21920.500000  149.300000   63.000000   23.300000
50%    252.000000  48057.000000  175.300000   67.000000   27.200000
75%    377.500000  75679.000000  199.500000   70.000000   31.750000
max    503.000000  99701.000000  255.900000   79.000000   37.700000

In [19]:
adverse_reactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   given_name        34 non-null     object
 1   surname           34 non-null     object
 2   adverse_reaction  34 non-null     object
dtypes: object(3)
memory usage: 944.0+ bytes


In [28]:
adverse_reactions.describe()

given_name  surname adverse_reaction
count          34       34               34
unique         34       33                6
top        steven  johnson     hypoglycemia
freq            1        2               19

In [14]:
adverse_reactions.adverse_reaction.value_counts()

hypoglycemia                 19
injection site discomfort     6
headache                      3
nausea                        2
throat irritation             2
cough                         2
Name: adverse_reaction, dtype: int64

Try selecting the records in the `patients` table for patients that are from the *city* New York.

In [17]:
len(patients[patients.city=='New York'])

18

In [31]:
patients[patients['address'].isnull()].count().patient_id

12

In [36]:
##look at patients surnames
patients.surname.value_counts()

Doe                  6
Jakobsen             3
Taylor               3
Lund                 2
Souza                2
Correia              2
Bùi                  2
Dratchev             2
Nilsen               2
Berg                 2
Grímsdóttir          2
Hueber               2
Tạ                   2
Cindrić              2
Lương                2
Ogochukwu            2
Liễu                 2
Batukayev            2
Parker               2
Lâm                  2
Collins              2
Kowalczyk            2
Johnson              2
Gersten              2
Silva                2
Schiavone            2
Aranda               2
Tucker               2
Cabrera              2
Kadyrov              2
Woźniak              2
Fejes                1
Flamand              1
Rukavina             1
Chidi                1
Hrdá                 1
Carvalho             1
Henzen               1
Rosario              1
Prince               1
Wolfe                1
Nowakowski           1
Borgen               1
Stanić     

In [37]:
patients.address.value_counts()

123 Main Street                  6
648 Old Dear Lane                2
2778 North Avenue                2
2476 Fulton Street               2
922 Chapmans Lane                1
3548 Hilltop Drive               1
3411 Pyramid Valley Road         1
2152 Heritage Road               1
4932 Goldleaf Lane               1
4851 Andy Street                 1
2704 Windy Ridge Road            1
954 Summit Park Avenue           1
2389 Rubaiyat Road               1
479 Elmwood Avenue               1
4346 Sussex Court                1
4476 Center Street               1
4941 Marion Drive                1
4567 Hazelwood Avenue            1
1589 Roosevelt Wilson Lane       1
3106 Evergreen Lane              1
3261 Desert Broom Court          1
2438 Shady Pines Drive           1
3539 Bottom Lane                 1
4380 Riverside Drive             1
350 Ross Street                  1
108 Griffin Street               1
664 Lyon Avenue                  1
3418 Tyler Avenue                1
34 Hamill Avenue    

In [38]:
patients[patients.address.duplicated()]

patient_id assigned_sex given_name     surname             address         city     state  zip_code        country                                     contact   birthdate  weight  height   bmi
29           30         male       Jake    Jakobsen   648 Old Dear Lane  Port Jervis  New York   12771.0  United States  JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67  24.4
219         220         male         Mỹ       Quynh                 NaN          NaN       NaN       NaN            NaN                                         NaN    4/9/1978   237.8      69  35.1
229         230         male       John         Doe     123 Main Street     New York        NY   12345.0  United States                 johndoe@email.com1234567890    1/1/1975   180.0      72  24.4
230         231       female  Elisabeth     Knudsen                 NaN          NaN       NaN       NaN            NaN                                         NaN   9/23/1976   165.9      63  29.4
234         235       female    Martina   Tománková                 NaN          NaN       NaN       NaN            NaN                                         NaN    4/7/1936   199.5      65  33.2
237         238         male       John         Doe     123 Main Street     New York        NY   12345.0  United States                 johndoe@email.com1234567890    1/1/1975   180.0      72  24.4
242         243         male       John     O'Brian                 NaN          NaN       NaN       NaN            NaN                                         NaN   2/25/1957   205.3      74  26.4
244         245         male       John         Doe     123 Main Street     New York        NY   12345.0  United States                 johndoe@email.com1234567890    1/1/1975   180.0      72  24.4
249         250         male   Benjamin      Mehler                 NaN          NaN       NaN       NaN            NaN                                         NaN  10/30/1951   146.5      69  21.6
251         252         male       John         Doe     123 Main Street     New York        NY   12345.0  United States                 johndoe@email.com1234567890    1/1/1975   180.0      72  24.4
257         258         male        Jin        Kung                 NaN          NaN       NaN       NaN            NaN                                         NaN   5/17/1995   231.7      69  34.2
264         265       female   Wafiyyah      Asfour                 NaN          NaN       NaN       NaN            NaN                                         NaN   11/3/1989   158.6      63  28.1
269         270       female     Flavia  Fiorentino                 NaN          NaN       NaN       NaN            NaN                                         NaN   10/9/1937   175.2      61  33.1
277         278         male       John         Doe     123 Main Street     New York        NY   12345.0  United States                 johndoe@email.com1234567890    1/1/1975   180.0      72  24.4
278         279       female   Generosa       Cabán                 NaN          NaN       NaN       NaN            NaN                                         NaN  12/16/1962   124.3      69  18.4
282         283       female      Sandy      Taylor  2476 Fulton Street     Rainelle        WV   25962.0  United States        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64  35.4
286         287         male      Lewis        Webb                 NaN          NaN       NaN       NaN            NaN                                         NaN    4/1/1979   155.3      68  23.6
296         297       female        Chỉ         Lâm                 NaN          NaN       NaN       NaN            NaN                                         NaN   5/14/1990   181.1      63  32.1
502         503         male        Pat     Gersten   2778 North Avenue         Burr  Nebraska   68324.0  United States        PatrickGersten@rhyta.com402-848-4923    5/3/1954   138.2      71  19.3

In [44]:
patients.weight.sort_values()

210     48.8
459    102.1
335    102.7
74     103.2
317    106.0
       ...  
144    244.9
61     244.9
283    245.5
118    254.5
485    255.9
Name: weight, Length: 503, dtype: float64

In [40]:
sum(treatments.auralin.isnull())

0

In [41]:
sum(treatments.novodra.isnull())

0